k9mail/src/main/java/com/fsck/k9/ui/crypto/MessageCryptoHelper.java

```diff
@@ -579,8 +583,17 @@ private void addCryptoResultAnnotationToMessage(CryptoResultAnnotation resultAnn
     }
 
     private void onCryptoFinished() {
-        currentCryptoPart = null;
-        partsToDecryptOrVerify.removeFirst();
+        boolean currentPartIsFirstInQueue = partsToDecryptOrVerify.peekFirst() == currentCryptoPart;
+        if (!currentPartIsFirstInQueue) {
+            throw new IllegalStateException(
+                    "Trying to remove part from queue that is not the currently processed one!");
+        }
+        if (currentCryptoPart != null) {
```

Comment:
If both the queue and `currentCryptoPart` are null, we don't throw but just skip (that's what happened in #1891 I hope). The check is to make sure we don't pop an element from the queue that is different from the one we have in `currentCryptoPart`, which could lead to very difficult to trace bugs down the road.

The PgpApi should never ever return more than one result, if it does we're in trouble.